metadata-io/src/main/java/com/linkedin/metadata/entity/DeleteEntityService.java

```diff
@@ -0,0 +1,254 @@
+@RequiredArgsConstructor
+public class DeleteEntityService {
+
+  private final EntityService _entityService;
+  private final GraphService _graphService;
+
+  private static final Integer ELASTIC_BATCH_DELETE_SLEEP_SEC = 5;
+
+  /**
+   * Public endpoint that deletes references to a given urn across DataHub's metadata graph. This is the entrypoint for
+   * addressing dangling pointers whenever a user deletes some entity.
+   *
+   * @param urn     The urn for which to delete references in DataHub's metadata graph.
+   * @param dryRun  Specifies if the delete logic should be executed to conclusion or if the caller simply wants a
+   *                preview of the response.
+   * @return  A {@link DeleteReferencesResponse} instance detailing the response of deleting references to the provided
+   * urn.
+   */
+  public DeleteReferencesResponse deleteReferencesTo(final Urn urn, final boolean dryRun) {
+    final DeleteReferencesResponse result = new DeleteReferencesResponse();
+    RelatedEntitiesResult relatedEntities =
+        _graphService.findRelatedEntities(null, newFilter("urn", urn.toString()), null,
+            EMPTY_FILTER,
+            ImmutableList.of(),
+            newRelationshipFilter(EMPTY_FILTER, RelationshipDirection.INCOMING), 0, 10000);
+
+    final List<RelatedAspect> relatedAspects = relatedEntities.getEntities().stream().map(relatedEntity -> {
+          final Urn relatedUrn = UrnUtils.getUrn(relatedEntity.getUrn());
+          final String relationType = relatedEntity.getRelationshipType();
+          final String relatedEntityName = relatedUrn.getEntityType();
+          final EntitySpec relatedEntitySpec = _entityService.getEntityRegistry().getEntitySpec(relatedEntityName);
+
+          return getAspectReferringTo(urn, relatedUrn, relationType, relatedEntitySpec).map(pair -> {
+            final RelatedAspect relatedAspect = new RelatedAspect();
+            relatedAspect.setEntity(relatedUrn);
+            relatedAspect.setRelationship(relationType);
+            relatedAspect.setAspect(pair.getSecond().getName());
+            return relatedAspect;
+          });
+        }).filter(java.util.Optional::isPresent)
+        .map(java.util.Optional::get)
+        .limit(10).collect(Collectors.toList());
+
+    result.setRelatedAspects(new RelatedAspectArray(relatedAspects));
+    result.setTotal(relatedEntities.getTotal());
+
+    if (dryRun) {
+      return result;
+    }
+
+    int processedEntities = 0;
+
+    // Delete first 10k
+    relatedEntities.getEntities().forEach(entity -> deleteReference(urn, entity));
+
+    processedEntities += relatedEntities.getEntities().size();
+
+    // Delete until less than 10k are left
+    while (relatedEntities.getCount() > 10000) {
+      log.info("Processing batch {} of {} aspects", processedEntities, relatedEntities.getTotal());
+      sleep(ELASTIC_BATCH_DELETE_SLEEP_SEC);
+      relatedEntities = _graphService.findRelatedEntities(null, newFilter("urn", urn.toString()),
```

Comment:
What if we fail to find related entities? any try catching we need to think about?

Reply:
If there are related aspect instances that we do not find then that is a graph index problem. In that case this whole logic breaks as that is one of the cores requirements.

Reply:
I'd like to see a loud and clear place where we can know if something fails and why. That's my ask!